### PR TITLE
Drop python3.8 support core and providers

### PR DESCRIPTION
**Pull request by jscheffl**
As Python 3.8 is getting out-of support, this PR removes support from Airflow providers and code.

See Python release schedule: https://peps.python.org/pep-0569/

Continuation of PR #42742 from apache repo

### CODE REVIEW

**Comment by jscheffl** [comment]
So.. green.Now we can wait until 2024-10-14, rebase and merge...

**Comment by potiuk** [comment]
> So.. green.Now we can wait until 2024-10-14, rebase and merge...

Wooohooo. Let me also review more closely and look at the jobs running.
I am a bit suspicious on things like that "just working" - I guess there is a catch somewhere :)

File: .github/workflows/check-providers.yml
Comment (potiuk) [inline .github/workflows/check-providers.yml:113]: There is a bug two lines above - `${{ matrix.python-version }}` is used - should be 3.9 explicitly.

File: .pre-commit-config.yaml
Comment (potiuk) [inline .pre-commit-config.yaml:1069]: Nice

File: README.md
Comment (potiuk) [inline README.md]: 2.10.2 still uses "3.8" min version - and the whole 2.* line should continue doing so.

File: airflow/providers/amazon/aws/transfers/sql_to_s3.py
Comment (potiuk) [inline airflow/providers/amazon/aws/transfers/sql_to_s3.py:226]: Related?

File: dev/breeze/src/airflow_breeze/global_constants.py
Comment (potiuk) [inline dev/breeze/src/airflow_breeze/global_constants.py:26]: Finally!

File: dev/breeze/src/airflow_breeze/utils/cdxgen.py
Comment (potiuk) [inline dev/breeze/src/airflow_breeze/utils/cdxgen.py]: We should leave those in - we use it to generate SBOMs for old versions of Airlfow and we need all historical versions of Python in this mapping.

File: docs/docker-stack/changelog.rst
Comment (potiuk) [inline docs/docker-stack/changelog.rst]: Well. So far we only had approach that we remove python version for the next M|NOR version - so 2.10.3 will still have to support version 3.8

File: generated/PYPI_README.md
Comment (potiuk) [inline generated/PYPI_README.md]: Same here 2.10.2 still aupports Python 3.8 and 2.10.3 still should IMHO.

**Review by potiuk — COMMENTED** [review]
Few comments - mostly about support for Python 3.8 in 2.10 (and potentially 2.11) - I think according to our rules 2.10.3+ should still support Python 3.8 (because also we promise "smoothest possible" migration in patchlevel releases. 

Likely 2.11.0 should also be Python 3.8 compliant - but this is something we can discuss later, we might very well tell our users that if they want to migrate to Airflow 3 using 2.11 that they should migrate their Python version to 3.9+ if they want to move to Airflow 3 via 2.11.

File: README.md
Comment (jscheffl) [reply README.md:102]: So no backport and dropping support in 2.10.3? Will this contiunue to support 3.8?

File: generated/PYPI_README.md
Comment (jscheffl) [reply generated/PYPI_README.md:59]: Aaah, then I mis-interpreted the chapter in docs/apache-airflow/installation/supported-versions.rst:55,,, Minor/Major drop the Py versions but patch release keep them... will 2.11 be an exception as we mentally keep the 2.10 line or would be drop it in 2.11 as well?

**Comment by jscheffl** [comment]
> Few comments - mostly about support for Python 3.8 in 2.10 (and potentially 2.11) - I think according to our rules 2.10.3+ should still support Python 3.8 (because also we promise "smoothest possible" migration in patchlevel releases.
> 
> Likely 2.11.0 should also be Python 3.8 compliant - but this is something we can discuss later, we might very well tell our users that if they want to migrate to Airflow 3 using 2.11 that they should migrate their Python version to 3.9+ if they want to move to Airflow 3 via 2.11.

...but this also means that users might get an update of core but no update of providers when staying with Python 3.8 post merging? (As we are not maintaining minor versions of providers in separate branches...)

File: dev/breeze/src/airflow_breeze/utils/cdxgen.py
Comment (jscheffl) [reply dev/breeze/src/airflow_breeze/utils/cdxgen.py]: I will restore but... are you sure? This is in pydoc... not in active code :-D

File: airflow/providers/amazon/aws/transfers/sql_to_s3.py
Comment (jscheffl) [reply airflow/providers/amazon/aws/transfers/sql_to_s3.py:226]: Yes, someow it seems Python 3.9 improved typing, such `mypy-providers` complained with:
```
airflow/providers/amazon/aws/transfers/sql_to_s3.py:226: error: Redundant cast
to "DataFrame"  [redundant-cast]
                    cast(
                    ^
```

File: docs/docker-stack/changelog.rst
Comment (jscheffl) [reply docs/docker-stack/changelog.rst]: Reverted this.

File: README.md
Comment (jscheffl) [reply README.md]: Reverted.

File: .github/workflows/check-providers.yml
Comment (jscheffl) [reply .github/workflows/check-providers.yml:113]: Adjusted.

File: .github/workflows/build-images.yml
Comment (ashb) [inline .github/workflows/build-images.yml]: Nit: we should probably start quoting these now, so that `3.10` doesn't become `3.1` (`x: 3.10` vs `x: "3.10"`) :grin:

**Review by kaxil — APPROVED** [review]

**Comment by jscheffl** [comment]
> > So.. green.Now we can wait until 2024-10-14, rebase and merge...
> 
> Wooohooo. Let me also review more closely and look at the jobs running. I am a bit suspicious on things like that "just working" - I guess there is a catch somewhere :)

Yeah, this requires proper review - such a large change set there will be probably still some glitch be hidden... tried my best in multiple ways to look-up... hunt for the secret!!

File: .github/workflows/build-images.yml
Comment (jscheffl) [reply .github/workflows/build-images.yml]: Good point. Typical pitfall. On the other hand the `- name: "Setup python"` is not needing any quotes, this is a safe string :-D

Will wait for a full review cycle and then adjust as a bulk.

File: dev/breeze/src/airflow_breeze/utils/cdxgen.py
Comment (potiuk) [reply dev/breeze/src/airflow_breeze/utils/cdxgen.py]: Ah right 🤯

Comment (potiuk) [reply dev/breeze/src/airflow_breeze/utils/cdxgen.py]: You can remove it - this part is anyhow something that we will have to improve a bit later  - when we improve SBOM generation (there are few things to do in the future).

File: .github/workflows/build-images.yml
Comment (potiuk) [reply .github/workflows/build-images.yml]: Indeed. Great catch @ashb !

File: dev/breeze/doc/01_installation.rst
Comment (shahar1) [inline dev/breeze/doc/01_installation.rst]:
```suggestion
    If you used Breeze with Python 3.8 and when running it, it will complain that it needs Python 3.9. In this
```

File: contributing-docs/03_contributors_quick_start.rst
Comment (shahar1) [inline contributing-docs/03_contributors_quick_start.rst]: FWIW 😅

```suggestion
   platform linux -- Python 3.9.20, pytest-8.3.3, pluggy-1.5.0 -- /usr/local/bin/python
```

You could change the other references to 3.9 subversions in the other files if you want, I assume that it's not too critical

File: dev/breeze/doc/adr/0002-implement-standalone-python-command.md
Comment (shahar1) [inline dev/breeze/doc/adr/0002-implement-standalone-python-command.md:141]:
```suggestion
   * Python 3.9+ (Python 3.9 end-of-life is in October 2025)
```

File: dev/breeze/src/airflow_breeze/global_constants.py
Comment (shahar1) [inline dev/breeze/src/airflow_breeze/global_constants.py:324]: Good catch

File: docs/apache-airflow/installation/supported-versions.rst
Comment (shahar1) [inline docs/apache-airflow/installation/supported-versions.rst]:
```suggestion
   are Python 3.8 images, however, in the first MINOR/MAJOR release of Airflow released after October 14, 2024.
```
(or other datetime formatting acceptable on both US and the rest of the world 😀)

Also, are we ok with this specific date?

**Review by shahar1 — APPROVED** [review]
Looks great!
Some minor nits, not blocking for merging though :)

File: .github/workflows/build-images.yml
Comment (jscheffl) [reply .github/workflows/build-images.yml]: Fixed

File: dev/breeze/doc/01_installation.rst
Comment (jscheffl) [reply dev/breeze/doc/01_installation.rst]: Good catch!

File: contributing-docs/03_contributors_quick_start.rst
Comment (jscheffl) [reply contributing-docs/03_contributors_quick_start.rst]: Updated the full pytest logs :-D

File: dev/breeze/doc/adr/0002-implement-standalone-python-command.md
Comment (jscheffl) [reply dev/breeze/doc/adr/0002-implement-standalone-python-command.md:141]: nah, I don't want to touch old ADR docs, they stay as they are once generated.

**Comment by jscheffl** [comment]
If CI is green... LGTM? I am going to bed and leave this for the "late shift" == US time zone :-D

**Comment by uranusjr** [comment]
I don't think 2.11 needs to support 3.8. The release is a bridge to 3.0, and there’s no point upgrading to it if you intend to stay on 3.8. Conversely, if you want to upgrade to 3.0, you must upgrade Python first, so it’s reasonable to expect users to do that first before upgrading to 2.11.

File: docs/apache-airflow/installation/supported-versions.rst
Comment (ashb) [inline docs/apache-airflow/installation/supported-versions.rst]:
```suggestion
   are Python 3.8 images, however, in the first MINOR/MAJOR release of Airflow released after 2024-10-14,
```

(because spelling is complaining about `th` and this is an easier fix from GH review.)

**Comment by potiuk** [comment]
>  I don't think 2.11 needs to support 3.8. The release is a bridge to 3.0, and there’s no point upgrading to it if you intend to stay on 3.8. Conversely, if you want to upgrade to 3.0, you must upgrade Python first, so it’s reasonable to expect users to do that first before upgrading to 2.11.

Agreed. The 2.10.* should support 3.8 but 2.11 should not. Which means that we should cherry-pick that one to the v2-test branch when we prepare for 2.11 release - in the meantime 2.10.3/4 will be released from that branch with 3.8 support, so we should not cherry-pick it **yet**.

**Comment by potiuk** [comment]
I updated milestone to 2.11.0.

**Comment by jscheffl** [comment]
> Agreed. The 2.10.* should support 3.8 but 2.11 should not. Which means that we should cherry-pick that one to the v2-test branch when we prepare for 2.11 release - in the meantime 2.10.3/4 will be released from that branch with 3.8 support, so we should not cherry-pick it **yet**.

No "easy" cherry-pick as main has diverged in many details already. I started a v2-10 backport PR already in https://github.com/apache/airflow/pull/42788 and it took ~3h of my lifetime to merge and find differences.

Branch was deleted but from PR/commit can be created again and made green.

Let's discuss tomorrow... after the initial removal I'd assume it takes another couple of hours again once we cut v2.11, so we should have an agreement on removal not to waste efforts... (again :-D)

**Comment by kaxil** [comment]
Hopefully those changes are small, but some of these might be unavoidable I think

**Pull request merged**
